python/cuvs/cuvs/neighbors/ivf_pq/ivf_pq.pyx

```diff
@@ -0,0 +1,430 @@
+
+    cdef Index idx = Index()
+    cdef cuvsError_t build_status
+    cdef cydlpack.DLManagedTensor* dataset_dlpack = \
+        cydlpack.dlpack_c(dataset_ai)
+    cdef cuvsIvfPqIndexParams* params = index_params.params
+
+    cdef cuvsResources_t res = <cuvsResources_t>resources.get_c_obj()
+
+    with cuda_interruptible():
+        check_cuvs(cuvsIvfPqBuild(
+            res,
+            params,
+            dataset_dlpack,
+            idx.index
+        ))
+        idx.trained = True
+
+    return idx
+
+
+cdef _map_dtype_np_to_cuda(dtype, supported_dtypes=None):
+    if supported_dtypes is not None and dtype not in supported_dtypes:
+        raise TypeError("Type %s is not supported" % str(dtype))
+    return {np.float32: cudaDataType_t.CUDA_R_32F,
+            np.float16: cudaDataType_t.CUDA_R_16F,
+            np.uint8: cudaDataType_t.CUDA_R_8U}[dtype]
+
+
+cdef class SearchParams:
+    """
+    Supplemental parameters to search IVF-Pq index
+
+    Parameters
+    ----------
+    n_probes: int
+        The number of clusters to search.
+    lut_dtype: default = np.float32
+        Data type of look up table to be created dynamically at search
+        time. The use of low-precision types reduces the amount of shared
+        memory required at search time, so fast shared memory kernels can
+        be used even for datasets with large dimansionality. Note that
+        the recall is slightly degraded when low-precision type is
+        selected. Possible values [np.float32, np.float16, np.uint8]
+    internal_distance_dtype: default = np.float32
+        Storage data type for distance/similarity computation.
+        Possible values [np.float32, np.float16]
+    """
+
+    cdef cuvsIvfPqSearchParams* params
+
+    def __cinit__(self):
+        cuvsIvfPqSearchParamsCreate(&self.params)
+
+    def __dealloc__(self):
+        check_cuvs(cuvsIvfPqSearchParamsDestroy(self.params))
+
+    def __init__(self, *, n_probes=20, lut_dtype=np.float32,
+                 internal_distance_dtype=np.float32):
+        self.params.n_probes = n_probes
+        self.params.lut_dtype = _map_dtype_np_to_cuda(lut_dtype)
+        self.params.internal_distance_dtype = \
+            _map_dtype_np_to_cuda(internal_distance_dtype)
+
+    @property
+    def n_probes(self):
+        return self.params.n_probes
+
+    @property
+    def n_probes(self):
+        return self.params.n_probes
+
+    @property
+    def lut_dtype(self):
+        return self.params.lut_dtype
+
+    @property
+    def internal_distance_dtype(self):
+        return self.params.internal_distance_dtype
+
+
+@auto_sync_resources
+@auto_convert_output
+def search(SearchParams search_params,
+           Index index,
+           queries,
+           k,
+           neighbors=None,
+           distances=None,
+           resources=None):
+    """
+    Find the k nearest neighbors for each query.
+
+    Parameters
+    ----------
+    search_params : SearchParams
```

Comment:
Does this automatically to the ivf_pq.SearchParas in the docs? If not, we might want to fully-qualify these

Reply:
good point about the docs! We were missing the RST files for ivf-pq and bfknn, so it didn't show up at all =( - I've added those in.

For linking to the SearchParams struct - I had to update to `:py:class:cuvs.neighbors.ivf_pq.SearchParams` (like both add the fully qualified class name, and the `py:class` marker to create the hyper link) - but with that change we get a clickable link in the sphinx docs